File: OptimalPlacement.java
//time - O(HW C n) as in nCr
// space - O(HW)
class optimalPlacement {
    
    int result = Integer.MAX_VALUE; //return value
    
    public int bfs(int[][] board) {
        //same as 01 matrix
        int maxDistance = 0; //max distance between a parking lot(-1) and building(0)
        Queue<int[]> support = new LinkedList<>();
        //cant manipulate board[][] as it comes from backtrack() - so create a temp[][] as a copy of board
        int[][] temp = new int[board.length][board[0].length];
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board[0].length; j++)
            {
                //process all 0s together, and replace -1s with infinity
                if(board[i][j] == 0)
                {
                    temp[i][j] = 0;
                    support.offer(new int[]{i, j});
                }
                else
                {
                    temp[i][j] = Integer.MAX_VALUE;
                }
            }
        }
        
        int[][] dirs = {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
        while(!support.isEmpty())
        {
            int[] current = support.poll();
            for(int[] dir : dirs)
            {
                //for each neighbor, relax edge if a shorter distance is found
                int nRow = current[0] + dir[0];
                int nCol = current[1] + dir[1];
                if(nRow >= 0 && nRow < temp.length && nCol >= 0 && nCol < temp[0].length)
                {
                    if(temp[nRow][nCol] > 1 + temp[current[0]][current[1]])
                    {
                        temp[nRow][nCol] = 1 + temp[current[0]][current[1]];
                        support.offer(new int[]{nRow, nCol});
                        maxDistance = Math.max(maxDistance, temp[nRow][nCol]);
                    }
                }
            }
        }
        
        return maxDistance; //return max distance
    }
    
    public void backtrack(int[][] board, int buildings, int currentRow, int currentCol) {
        //base
        if(buildings == 0) //all buildings placed
        {
            int currentVal = bfs(board); //do dfs and update result if needed
            result = Math.min(result, currentVal);
            return;
        }
        //logic
        if(currentCol >= board[0].length)
        {
            currentRow++;
            currentCol = 0;
        }
        for(int i = currentRow; i < board.length; i++)
        {
            for(int j = currentCol; j < board[0].length; j++)
            {
                board[i][j] = 0; //place building at (i,j)
                backtrack(board, buildings - 1, i, j + 1); //recurse & place building at same row next col
                board[i][j] = -1; //bactrack to explore other possiblities
            }
        }
        return;
    }
    
    public int findMin(int H, int W, int n) {
        int[][] board = new int[H][W];
        for(int i = 0; i < board.length; i++)
        {
            for(int j = 0; j < board[0].length; j++)
            { 
                board[i][j] = -1; //fill each cell with -1s
            }
        }
        backtrack(board, n, 0, 0); //start placing from cell 0,0
        return result;
    }
    
    public static void main(String[] args) {
        optimalPlacement obj = new optimalPlacement();
        System.out.println(obj.findMin(4, 4, 1));
    }
}

File: OptimalPlacement.py
class OptimalPlacement:
    def __init__(self, height, width, buildings):
        self.height = height
        self.width = width
        self.number = buildings
        self.output = float("inf")

    def find_min_distance(self):
        grid = [[-1]*self.width for _ in range(self.height)]
        self._backtrack(grid, 0,0, self.number)
        return self.output
    def _backtrack(self, grid, r, c, n):
        
        if n == 0:
            # if self._bfs(grid)>20:
            #     print("yoyoyo")
            #     print(self._bfs(grid))
            
            num = self._bfs(grid[:])
            # print("Passing grid as",grid,n,num )
            self.output = min(self.output, num)
            return 
        if c == self.width:
            r+=1
            c =0
        
        for i in range(r,self.height):
            for j in range(c, self.width):
                #action
                grid[i][j] = 1
                #recurse

                self._backtrack(grid, i, j+1, n-1)                    
                #backtrack
                grid[i][j] = -1
            c = 0

    def _bfs(self, grid):
        
        queue = deque([[i,j] for i in range(len(grid)) for j in range(len(grid[0])) if grid[i][j] == 1])        

        # print("Length of queue",len(queue))
        distance = 0
        visited = set()
        while queue:
            for _ in range(len(queue)):
                a,b = queue.popleft()
                visited.add((a,b))
                for i,j in [[a+1,b],[a-1,b],[a,b+1],[a,b-1]]:
                    if 0<=i<self.height and 0<=j<self.width and grid[i][j] == -1 and (i,j) not in visited:
                        # print("Adding", i,j)
                        queue.append([i,j])
                        visited.add((i,j))
            distance+=1
        
        return distance-1
                    

def main():
    p = OptimalPlacement(4, 4, 3)
    print(p.find_min_distance())


main()
'''
# TIME COMPLEXITY: O(height*width C buildings) - nCr -> Asymtotically, exponential
# SPACE COMPLEXITY: O(height*width)
